Make Timestamps Returned by Snapshot APIs Consistent

original-brownbear opened the pull request:
* We don't have to calculate the start and end times form the shards for the status API, we have the start time available from the CS or the `SnapshotInfo` in the repo and can either take the end time form the `SnapshotInfo` or
take the most recent time from the shard stats for in progress snapshots
* Had to add serialization of these time values since we can't compute them from the shards anymore
   * Fall back to computing time values from shards in BwC scenario 
* Started a new IT class for this since the existing ones are huge already and we have a few more API inconsistencies with snapshots to fix and add tests for
* Closes #43074

------------

Admittedly the weird mutable data structures, `SnapshotStats` in particular becomes a little more magical than it already is with this change. I'd try to clean this up in a follow-up.

original-brownbear committed:
Make Timestamps returned by Snapshot APIs Consistent

* We don't have to calculate the start and end times form the shards for the status API, we have the start time available from the CS or the `SnapshotInfo` in the repo and can either take the end time form the `SnapshotInfo` or
take the most recent time from the shard stats for in progress snapshots
* Closes #43074

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear commented:
Jenkins run elasticsearch-ci/2

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotStats.java:
```diff
@@ -46,6 +46,11 @@
     SnapshotStats() {
     }
 
+    SnapshotStats(long startTime, long time) {
```
I don't think we need another constructor just for this, maybe reuse the existing one?

original-brownbear replied:
Sure, I guess I hated putting all those `0` but it's shorted this way for sure and horrible regardless :D

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotStatus.java:
```diff
@@ -71,14 +72,14 @@
         this.includeGlobalState = includeGlobalState;
         shardsStats = new SnapshotShardsStats(shards);
-        updateShardStats();
+        updateShardStats(startTime, time);
```
Can we check that they are not < 0?

original-brownbear replied:
Actually, I don't think we need to I think :) We get these numbers from `org.elasticsearch.index.snapshots.IndexShardSnapshotStatus` which does all the checks internally so we should be good here?

tlrx replied:
:+1:

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java:
```diff
@@ -239,8 +239,10 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
                     builder.add(new SnapshotStatus(new Snapshot(repositoryName, snapshotId), state,
-                        Collections.unmodifiableList(shardStatusBuilder), snapshotInfo.includeGlobalState()));
+                        Collections.unmodifiableList(shardStatusBuilder), snapshotInfo.includeGlobalState(),
+                        startTime, snapshotInfo.endTime() - startTime));
```
I don't think that SnapshotInfo guarantees that endTime > startTime, maybe we should be defensive here?

original-brownbear replied:
Well no it doesn't but this isn't a new issue is it?  It just shows up in two places now instead of in one. If you have skewed clocks and a master failover during snapshotting the absolute timestamp stored in the CS won't work out. All that happens now is that we see the same issue in two places instead of one?

tlrx replied:
Ok

tlrx commented on server/src/test/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java:
```diff
@@ -0,0 +1,76 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
```
I don't think this is needed, the default scope SUITE should be fine

original-brownbear replied:
I guess for a single test it doesn't matter anyway, but in general I'm not a big fan of the suit scope when running tests that so heavily depend on the exact state of the cluster like here -> that's why I went with test scope. Not sure if that makes sense?

tlrx replied:
I still don't think there is a reason here to use TEST and not the default, unless you noticed test failures with the SUITE test scope and if so we need to investigate them.

original-brownbear replied:
Fine with me :) Removed it.

tlrx reviewed (approved):
LGTM, thanks for taking care of this. I left some minor comments

I agree that SnapshotStats would benefit from a clean up

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 43074

original-brownbear committed:
CR: remove redundant ctor

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 43074

original-brownbear commented:
After merging `master` into this branch (even after resolving the compile issue) this is actually failing tests. Looking into it, something is off now with the timestamps somewhere.

original-brownbear committed:
Fix logic to work with clock skew

original-brownbear committed:
fix in progress reported timestamp

original-brownbear commented:
@tlrx I have to admit, your initial instincts about the possibility of negative timestamps were correct after all. I had to push 6223466 after merging `master`. There were two issues here:
1. The code I changed from using relative to absolute time needed absolute time. Otherwise, it will not work in a BwC scenario (because in those we will still use the time stamp on the shards to calculate the end and start times) so I moved all the logic in the repo and snapshot handling that was relevant to the same absolute time method on the thread pool to get things consistent here
2. The logic of calculating the start and end time needs to be disabled when using timestamps from `SnapshotInfo`. This only worked fine when all the code was using `System.currentTimeMillis` but stopped working now with cached time. Since all the nodes in the test cluster use different thread pools that was simulating clock skew, which lead to the calculation not working out anymore (since due to skew a shard could look like it started before the snapshot itself or finished after).

Can you take another look please? :)

tlrx reviewed (approved):
LGTM. I left a comment about test scope which I think is important.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into 43074

original-brownbear committed:
CR: remove test scope

original-brownbear commented:
Jenkins run elasticsearch-ci/1
